File: assignment5/ClientTCP.java
import java.net.*;  // for DatagramSocket, DatagramPacket, and InetAddress
import java.io.*;   // for IOException
import java.util.Scanner; // for Scanner


public class ClientTCP {

   public static void main(String args[]) throws Exception {
   
      if (args.length != 2 && args.length != 3) // Test for correct # of args
         throw new IllegalArgumentException("Parameter(s): <Destination>" +
                                                   " <Port> [<encoding]");
   
   
      InetAddress destAddr = InetAddress.getByName(args[0]); // Destination address
      int destPort = Integer.parseInt(args[1]) + TCPRequestBinConst.GROUP_NUMBER; // Destination port
      Byte TML = 6;
      Byte ID = 1;
      Byte operands = 1;
      Byte opCode;
      Short op1 = 0;
      Short op2 = 0;
   
      for(;;) {
         
      
         Scanner scan = new Scanner(System.in);
         System.out.println("Please Enter The Following Values:");
         System.out.print("\tOperand Type - \n\t\"0\" = +, \n\t\"1\" = -, \n\t\"2\" = *, \n\t\"3\" = /, \n\t\"4\" = >>, \n\t\"5\" = <<, \n\t\"6\" = ~, \n\t\"exit\" = Terminate \n\tEnter Selection: ");
      
         String opString = scan.nextLine();
      
         if (opString.equals("exit")) {
            break;
         }
         int opCodeInt = Integer.parseInt(opString);
         opCode = (byte) opCodeInt;
      
      
      
         System.out.print("\n\tOperand 1: ");
         int op1Int = Integer.parseInt(scan.nextLine());
         op1 = (short) op1Int;
      
      
         if (opCode <= 5) {
            System.out.print("\n\tOperand 2: ");
            int op2Int = Integer.parseInt(scan.nextLine());
            op2 = (short) op2Int;
            TML = (byte) 8;
            operands = (byte) 2;
         }
      
         TCPRequest Request = new TCPRequest(TML, ID, opCode, operands, op1, op2);
      
         Socket sock = new Socket(destAddr, destPort); // TCP socket for sending
      
      
                // Use the encoding scheme given on the command line (args[2])
         TCPRequestEncoder encoder = (args.length == 3 ?
                                             new TCPRequestEncoderBin(args[2]) :
                                             new TCPRequestEncoderBin());
      
         byte[] codedRequest = encoder.encode(Request); // Encode Request
      
      
         System.out.println("\nMessage length: " + codedRequest.length);
         System.out.print("Request Hex String: \n");
         //byte[] requestBuffer = message.getData();
         for (int i = 0; i < Request.TML; i++) {
            System.out.format("\t0x%x\n", codedRequest[i]);
         }
         InputStream in = sock.getInputStream();
         OutputStream out = sock.getOutputStream();
      
      
         TCPRequestDecoder decoder = (args.length == 2 ? // Which encoding
                                             new TCPRequestDecoderBin(args[1]) :
                                             new TCPRequestDecoderBin() );
      
         out.write(codedRequest);
         long start = System.nanoTime();
      
                //TODO DETERMINE IF DECODE CAN BE REMOVED FROM THIS STATEMENT
                //  TO GET ACCURATE SYSTEM TIME
         TCPResponse Response = decoder.decodeResponse(in);
         long finish = System.nanoTime();
         long elapsed = finish - start;
      
      
                //TODO FIX!!!
         System.out.print("Response Hex String: \n");
         byte[] codedResponse = toByteArray(Response);
         for (int i = 0; i < 7; i++) {
            System.out.format("\t0x%x\n", codedResponse[i]);
         }
      
         System.out.println("\n" + Response);
         System.out.println("Time Elapsed: " + elapsed / 1000000.0 + "ms\n");
         ID++;
      }
      //sock.close();
   }

   public static byte[] toByteArray(TCPResponse Response) throws IOException {
   
      ByteArrayOutputStream os = new ByteArrayOutputStream();
      DataOutputStream out = new DataOutputStream(os);
   
      out.writeByte(Response.TML);
      out.writeByte(Response.ID);
      out.writeByte(Response.error);
      out.writeInt(Response.result);
      return os.toByteArray();
   }

}

File: assignment5/ClientTCP.py
import socket
import sys
import time

GROUP_NUMBER = 24
DEFAULT_ENCODING = 'iso-8859-1'
MAX_WIRE_LENGTH = 1024

ipAddress = sys.argv[1]
port = int(sys.argv[2])
destPort = port + GROUP_NUMBER
destAddr = (ipAddress, destPort)


if len(sys.argv) != 2 and len(sys.argv) != 3:
   raise ValueError('Parameter(s): <Destination>  <Port> [<encoding]')


ID = 1
while(True):
   TML = None
   operands = None
   opCode = None
   op1 = None
   op2 = None

   print("\nPlease Enter The Following Values:\n")

   try:
      opString = raw_input(
          'Operand Type - \n\"0\" = + \n\"1\" = - \n\"2\" = * \n\"3\" = / \n\"4\" = >>  \n\"5\" = << \n\"6\" = ~ \n\"exit\" = Terminate \n\n\tEnter Selection: ')
      if opString == "exit":
         break
      opCode = int(opString)
      if(opCode < 0 or opCode > 6):
         opCode = None
   except:
      pass

   # Enter first operand

   try:
      op1 = int(raw_input("\tOperand 1: "))
      if (op1 < -32768 or op1 > 32767):
         op1 = None
   except:
      pass

   if(opCode < 6):
      TML = 8
      operands = 2

      try:
         op2 = int(raw_input("\tOperand 2: "))
         if (op2 < -32768 or op1 > 32767):
            op2 = None
      except:
         pass
   else:
      TML = 6
      numberOperands = 1
      operand2 = 0

   request = str('x%sx%sx%sx%sx%sx%sx' %
      (TML, ID, opCode, operands, op1, op2))

   codedRequest = bytearray(request)

   print("\nRequest Hex Values:")
   printString = str('%s %s %s %s %s %s' %
      (hex(TML), hex(ID), hex(opCode), hex(operands),
      hex(op1), hex(op2)))
   print(printString)

   sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
   sock.connect(destAddr)
   send = sock.sendall(request)
   start = int(round(time.time() * 1000))

   recv = sock.recv(MAX_WIRE_LENGTH)
   elapsed = int(round(time.time() * 1000)) - start

   response = [str(format(ord(x), 'x')).decode('utf-8') for x in recv]
   response = str(response)
   responseStr = ''

   for c in response:
       if c == 'a' or if c == 'b' or if c == 'c' or if c == 'd' or if c == 'e' or if c == 'f' or
       c == '0' or c == '1' or c == '2' or c == '3' or c == '4' or c == '5' or c == '6' or c == '7' or c == '8' or c == '9'
       responseStr = responseStr + c
       if c == 't':
           responseStr = responseStr + '9'
        if c == 'n':
            responseStr = responseStr + 'a'

    resultHex = int(responseStr[-8:], 16)
    result = -(resultHex & 0x8000) | (resultHex & 0x7fff)

   print("\nResponse Hex Values: ")
   print("%s" % response)
   print("Time Elapsed: %dms" % (elapsed))
   print("RequestID: %i" % ID)
   print("Result: ")
   print("Result: %i" % result)
   ID += 1

sock.close()
